### ISSUE TITLE
false-positive `name-defined` regression with dataclasses and shadowing

### ISSUE THREAD

**Issue by The-Compiler** [python/mypy#12907]
**Bug Report**

This file:

```python
from dataclasses import dataclass
import collections


@dataclass
class Shadow:

    collections: collections.deque
    other: collections.defaultdict


print(Shadow(collections.deque(), collections.defaultdict()))
```

is perhaps a bit unorthodox, but runs fine, because the `collections: ...` doesn't actually shadow the module when defining `other`.

**Actual Behavior**

mypy v0.950 was happy with it, but mypy v0.960 claims that:

```
test.py:9: error: Name "collections.defaultdict" is not defined
```

**Introducing commit**

Bisected to 03901ef657daf6f477e6fe933c42ac270cf5034b ("Running dataclass transform in a later pass to fix crashes (#12762)", @JukkaL)

**Your Environment**

- Mypy version used: v0.960
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.10.4
- Operating system and version: Archlinux

**Comment by AlexWaygood** [comment]
Looks similar to #10640. But it's not the same, because #10640 reproduces on 0.950, whereas this doesn't. That _might_ be because #10640 involves a `from foo import bar` import, whereas this is an `import foo` import.

**Comment by JukkaL** [comment]
I suspect that the original behavior was accidental. Now dataclass behavior is similar to ordinary classes, since #12762 made dataclasses a little less special:
```py
import collections

class Shadow:
    collections: collections.deque
    other: collections.defaultdict  # Error here
```

This is still a regression. As a workaround, you can import `collections` with a different name:
```py
import collections as _collections

class Shadow:
    collections: _collections.deque
    other: _collections.defaultdict  # OK
```

I'm going to try to find a simple fix, but I think that we can live with this regression if there is no easy fix in sight.
